core/embed/rust/src/ui/component/text/layout.rs

```diff
@@ -118,87 +144,76 @@ impl TextLayout {
-        ops: &mut dyn Iterator<Item = Op<'o>>,
-        cursor: &mut Point,
-        sink: &mut dyn LayoutSink,
-    ) -> LayoutFit {
-        let init_cursor = *cursor;
-        let mut total_processed_chars = 0;
-
-        for op in ops {
-            match op {
-                Op::Color(color) => {
-                    self.style.text_color = color;
-                }
-                Op::Font(font) => {
-                    self.style.text_font = font;
-                }
-                Op::Text(text) => match self.layout_text(text, cursor, sink) {
-                    LayoutFit::Fitting {
-                        processed_chars, ..
-                    } => {
-                        total_processed_chars += processed_chars;
-                    }
-                    LayoutFit::OutOfBounds {
-                        processed_chars, ..
-                    } => {
-                        total_processed_chars += processed_chars;
-
-                        return LayoutFit::OutOfBounds {
-                            processed_chars: total_processed_chars,
-                            height: self.layout_height(init_cursor, *cursor),
-                        };
-                    }
-                },
-            }
-        }
-
-        LayoutFit::Fitting {
-            processed_chars: total_processed_chars,
-            height: self.layout_height(init_cursor, *cursor),
-        }
+    /// Y coordinate of the bottom of the available space/bounds
+    pub fn bottom_y(&self) -> i16 {
+        (self.bounds.y1 - self.padding_bottom).max(self.bounds.y0)
     }
 
+    /// Loop through the `text` and try to fit it on the current screen,
+    /// reporting events to `sink`, which may do something with them (e.g. draw
+    /// on screen).
     pub fn layout_text(
         &self,
         text: &str,
         cursor: &mut Point,
         sink: &mut dyn LayoutSink,
+        continues_from_prev_page: bool,
     ) -> LayoutFit {
         let init_cursor = *cursor;
-        let bottom = (self.bounds.y1 - self.padding_bottom).max(self.bounds.y0);
         let mut remaining_text = text;
 
         // Check if bounding box is high enough for at least one line.
-        if cursor.y > bottom {
+        if cursor.y > self.bottom_y() {
             sink.out_of_bounds();
             return LayoutFit::OutOfBounds {
                 processed_chars: 0,
                 height: 0,
             };
         }
 
+        // Draw the arrow icon if we are in the middle of a string
+        if continues_from_prev_page {
+            let x_offset = sink.prev_page_ellipsis(*cursor, self);
+            cursor.x += x_offset;
+        }
+
         while !remaining_text.is_empty() {
+            let is_last_line = cursor.y + self.style.text_font.line_height() > self.bottom_y();
+            let line_ending_space = if is_last_line {
+                // TODO: find out the icon width
+                let width = self.style.text_font.text_width(PREV_PAGE_ELLIPSIS);
```

Comment:
We're at the end of the page though?

Reply:
Fixed in [b6fb917](https://github.com/trezor/trezor-firmware/pull/2803/commits/b6fb917d3111decae1eebac6d2a267c02e6b715d)